clash-lib/src/Clash/Core/HasType.hs

```diff
@@ -0,0 +1,245 @@
+    IntLiteral _ -> intPrimTy
+    WordLiteral _ -> wordPrimTy
+    StringLiteral _ -> stringPrimTy
+    FloatLiteral _ -> floatPrimTy
+    DoubleLiteral _ -> doublePrimTy
+    CharLiteral _ -> charPrimTy
+    Int64Literal _ -> int64PrimTy
+    Word64Literal _ -> word64PrimTy
+    NaturalLiteral _ -> naturalPrimTy
+    ByteArrayLiteral _ -> byteArrayPrimTy
+
+instance HasType PrimInfo where
+  coreTypeOf pr =
+    case primMultiResult pr of
+      SingleResult -> primType pr
+      MultiResult
+        | let (primArgs, primResTy) = splitFunForallTy (primType pr)
+        , TyConApp tupTcNm tupArgs <- tyView primResTy
+        , Text.isInfixOf "GHC.Tuple.(" (nameOcc tupTcNm)
+        -> mkPolyFunTy primResTy (primArgs <> fmap Right tupArgs)
+
+        | otherwise
+        -> error "PrimInfo.coreTypeOf: MultiResult primitive without tuple type"
+
+instance HasType Type where
+  coreTypeOf = id
+
+instance HasType (Var a) where
+  coreTypeOf = varType
+
+class InferType a where
+  inferCoreTypeOf :: TyConMap -> a -> Type
+
+instance InferType Term where
+  inferCoreTypeOf tcm = go
+   where
+    go = \case
+      Var i -> coreTypeOf i
+      Data dc -> coreTypeOf dc
+      Literal l -> coreTypeOf l
+      Prim pr -> coreTypeOf pr
+      Lam i x -> mkFunTy (coreTypeOf i) (go x)
+      TyLam i x -> ForAllTy i (go x)
+
+      x@App{} ->
+        case collectArgs x of
+          (fun, args) -> applyTypeToArgs x tcm (go fun) args
+
+      x@TyApp{} ->
+        case collectArgs x of
+          (fun, args) -> applyTypeToArgs x tcm (go fun) args
+
+      Letrec _ x -> go x
+      Case _ _ [] -> undefinedTy
+      Case _ _ (x:_) -> go (snd x)
+      Cast _ _ a -> a
+      Tick _ x -> go x
+
+ -- | Get the result type of a polymorphic function given a list of arguments
+applyTypeToArgs
+  :: Term
+  -> TyConMap
+  -> Type
+  -> [Either Term Type]
+  -> Type
+applyTypeToArgs e m opTy args = go opTy args
+ where
+  go opTy' []               = opTy'
+  go opTy' (Right ty:args') = goTyArgs opTy' [ty] args'
+  go opTy' (Left _:args')   = case splitFunTy m opTy' of
+    Just (_,resTy) -> go resTy args'
+    _ -> error $ unlines ["applyTypeToArgs:"
+                         ,"Expression: " ++ showPpr e
+                         ,"Type: " ++ showPpr opTy
+                         ,"Args: " ++ unlines (map (either showPpr showPpr) args)
+                         ]
+
+  goTyArgs opTy' revTys (Right ty:args') = goTyArgs opTy' (ty:revTys) args'
+  goTyArgs opTy' revTys args'            = go (piResultTys m opTy' (reverse revTys)) args'
+
+-- | Like 'piResultTyMaybe', but errors out when a type application is not
+-- valid.
+--
+-- Do not iterate 'piResultTy', because it's inefficient to substitute one
+-- variable at a time; instead use 'piResultTys'
+piResultTy
+  :: HasCallStack
+  => TyConMap
+  -> Type
+  -> Type
+  -> Type
+piResultTy m ty arg = case piResultTyMaybe m ty arg of
+  Just res -> res
+  Nothing  -> pprPanic "piResultTy" (ppr ty <> line <> ppr arg)
```

Comment:
I wonder if this can just be replaced with `piResultTy tcm ty arg = piResultTys tcm ty [arg]` and the `piResultTyMaybe` function then removed

Reply:
Updated: needs to be checked by someone else

Reply:
I mean, if we don't use `piResultTyMaybe` anywhere any longer, we might as well remove it.

Reply:
We don't, it was more of a _maybe this is some perf trick that I'm just not seeing_